[SDK-49] [Android] [expo-dev-client] Crash on android API_22

### Minimal reproducible example

https://github.com/GabrielDierks/test-expo-android-22/

### Summary

When creating a new expo app with the expo-dev-client on the latest SDK 49 on android SDK 22 the app crashes.

## Steps:

1. Create new expo project
2. Add the `expo-dev-client` project
3. Install all dependencies and run `npx expo prebuild --clean && npx expo run:android -d`
4. Build succeeds!
5. App crashes on open, adb logcat log:
`E/SoLoader(23418): couldn't find DSO to load: libflipper.so caused by: dlopen failed: cannot locate symbol "BN_set_flags" referenced by "/data/app/com.gabrielkadmos.testyyyy-1/lib/arm64/libflipper.so"... result: 0
E/AndroidRuntime(23418): FATAL EXCEPTION: FlipperConnectionThread
E/AndroidRuntime(23418): Process: com.gabrielkadmos.testyyyy, PID: 23418
E/AndroidRuntime(23418): java.lang.NoClassDefFoundError: <clinit> failed for class com.facebook.flipper.android.EventBase; see exception in other thread
E/AndroidRuntime(23418): 	at com.facebook.flipper.android.FlipperThread.run(FlipperThread.java:25)
E/AndroidRuntime(23418): FATAL EXCEPTION: FlipperEventBaseThread
E/AndroidRuntime(23418): Process: com.gabrielkadmos.testyyyy, PID: 23418
E/AndroidRuntime(23418): java.lang.UnsatisfiedLinkError: couldn't find DSO to load: libflipper.so caused by: dlopen failed: cannot locate symbol "BN_set_flags" referenced by "/data/app/com.gabrielkadmos.testyyyy-1/lib/arm64/libflipper.so"... result: 0`

6. Build config plugin to disable Flipper (see example repo ./application-plugin.js)
7. App crashes again with the following logs
`E/AndroidRuntime(23733): FATAL EXCEPTION: main
E/AndroidRuntime(23733): Process: com.gabrielkadmos.testyyyy, PID: 23733
E/AndroidRuntime(23733): java.lang.RuntimeException: Unable to start activity ComponentInfo{com.gabrielkadmos.testyyyy/com.gabrielkadmos.testyyyy.MainActivity}: java.lang.NoSuchFieldException: accessFlags
E/AndroidRuntime(23733): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2325)
E/AndroidRuntime(23733): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2387)
E/AndroidRuntime(23733): 	at android.app.ActivityThread.access$800(ActivityThread.java:151)
E/AndroidRuntime(23733): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1303)
E/AndroidRuntime(23733): 	at android.os.Handler.dispatchMessage(Handler.java:102)
E/AndroidRuntime(23733): 	at android.os.Looper.loop(Looper.java:135)
E/AndroidRuntime(23733): 	at android.app.ActivityThread.main(ActivityThread.java:5254)
E/AndroidRuntime(23733): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(23733): 	at java.lang.reflect.Method.invoke(Method.java:372)
E/AndroidRuntime(23733): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903)
E/AndroidRuntime(23733): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698)
E/AndroidRuntime(23733): Caused by: java.lang.NoSuchFieldException: accessFlags
E/AndroidRuntime(23733): 	at java.lang.Class.getDeclaredField(Class.java:890)
E/AndroidRuntime(23733): 	at expo.modules.ReactActivityDelegateWrapper.onCreate(ReactActivityDelegateWrapper.kt:126)
E/AndroidRuntime(23733): 	at com.facebook.react.ReactActivity.onCreate(ReactActivity.java:46)
E/AndroidRuntime(23733): 	at com.gabrielkadmos.testyyyy.MainActivity.onCreate(MainActivity.java:20)
E/AndroidRuntime(23733): 	at android.app.Activity.performCreate(Activity.java:5990)
E/AndroidRuntime(23733): 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1106)
E/AndroidRuntime(23733): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2278)
E/AndroidRuntime(23733): 	... 10 more
W/ActivityManager(19092):   Force finishing activity 1 com.gabrielkadmos.testyyyy/expo.modules.devlauncher.launcher.errors.DevLauncherErrorActivity`

### Environment

  expo-env-info 1.0.5 environment info:
    System:
      OS: macOS 14.0
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 16.17.0 - ~/.nvm/versions/node/v16.17.0/bin/node
      Yarn: 1.22.18 - /usr/local/bin/yarn
      npm: 8.15.0 - ~/.nvm/versions/node/v16.17.0/bin/npm
    Managers:
      CocoaPods: 1.11.3 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.0, iOS 17.0, macOS 14.0, tvOS 17.0, watchOS 10.0
      Android SDK:
        API Levels: 30, 31, 32, 33, 34
        Build Tools: 29.0.2, 30.0.2, 30.0.3, 31.0.0, 32.0.0, 32.1.0, 33.0.0, 33.0.0, 34.0.0
        System Images: android-22 | Google APIs ARM 64 v8a, android-32 | Google APIs ARM 64 v8a, android-32 | Google Play ARM 64 v8a, android-33 | Google APIs ARM 64 v8a, android-34 | Google APIs ARM 64 v8a
    IDEs:
      Android Studio: 2022.2 AI-222.4459.24.2221.10121639
      Xcode: 15.0/15A240d - /usr/bin/xcodebuild
    npmPackages:
      expo: ~49.0.13 => 49.0.13
      react: 18.2.0 => 18.2.0
      react-native: 0.72.5 => 0.72.5
    npmGlobalPackages:
      eas-cli: 5.4.0
      expo-cli: 6.3.10
    Expo Workflow: managed

--- a/packages/expo-dev-launcher/CHANGELOG.md
+++ b/packages/expo-dev-launcher/CHANGELOG.md
@@ -8,6 +8,8 @@
 
 ### 🐛 Bug fixes
 
+- [Android] Fixed crash from accessing the `accessFlags` attribute on Android API 22. ([#25248](https://github.com/expo/expo/pull/25248) by [@kudo](https://github.com/kudo))
+
 ### 💡 Others
 
 ## 2.4.13 — 2023-09-25

--- a/packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/helpers/DevLauncherReflectionExtensions.kt
+++ b/packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/helpers/DevLauncherReflectionExtensions.kt
@@ -3,6 +3,7 @@ package expo.modules.devlauncher.helpers
 import java.lang.Exception
 import java.lang.reflect.Field
 import java.lang.reflect.Modifier
+import expo.modules.devmenu.helpers.removeFinalModifier
 
 fun <T> Class<T>.getFieldInClassHierarchy(fieldName: String): Field? {
   var currentClass: Class<*>? = this
@@ -19,15 +20,8 @@ fun <T> Class<T>.getFieldInClassHierarchy(fieldName: String): Field? {
 
 fun <T> Class<out T>.setProtectedDeclaredField(obj: T, filedName: String, newValue: Any, predicate: (Any?) -> Boolean = { true }) {
   val field = getDeclaredField(filedName)
-  val modifiersField = Field::class.java.getDeclaredField("accessFlags")
-
+  removeFinalModifier(field)
   field.isAccessible = true
-  modifiersField.isAccessible = true
-
-  modifiersField.setInt(
-    field,
-    field.modifiers and Modifier.FINAL.inv()
-  )
 
   if (!predicate.invoke(field.get(obj))) {
     return

--- a/packages/expo-dev-menu/CHANGELOG.md
+++ b/packages/expo-dev-menu/CHANGELOG.md
@@ -8,6 +8,8 @@
 
 ### 🐛 Bug fixes
 
+- [Android] Fixed crash from accessing the `accessFlags` attribute on Android API 22. ([#25248](https://github.com/expo/expo/pull/25248) by [@kudo](https://github.com/kudo))
+
 ### 💡 Others
 
 ## 3.2.1 — 2023-09-25

--- a/packages/expo-dev-menu/android/src/main/java/expo/modules/devmenu/helpers/DevMenuReflectionExtensions.kt
+++ b/packages/expo-dev-menu/android/src/main/java/expo/modules/devmenu/helpers/DevMenuReflectionExtensions.kt
@@ -1,5 +1,6 @@
 package expo.modules.devmenu.helpers
 
+import android.os.Build
 import java.lang.reflect.Field
 import java.lang.reflect.Modifier
 
@@ -12,15 +13,23 @@ fun <T, R> Class<out T>.getPrivateDeclaredFieldValue(filedName: String, obj: T):
 
 fun <T> Class<out T>.setPrivateDeclaredFieldValue(filedName: String, obj: T, newValue: Any) {
   val field = getDeclaredField(filedName)
-  val modifiersField = Field::class.java.getDeclaredField("accessFlags")
-
+  removeFinalModifier(field)
   field.isAccessible = true
-  modifiersField.isAccessible = true
-
-  modifiersField.setInt(
-    field,
-    field.modifiers and Modifier.FINAL.inv()
-  )
-
   field.set(obj, newValue)
+}
+
+@Suppress("DiscouragedPrivateApi")
+fun removeFinalModifier(field: Field) {
+  if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
+    val artField = Field::class.java.getDeclaredField("artField")
+    artField.isAccessible = true
+    val modifiers = Class.forName("java.lang.reflect.ArtField").getDeclaredField("accessFlags")
+    modifiers.isAccessible = true
+    modifiers.setInt(artField.get(field), field.modifiers and Modifier.FINAL.inv())
+    return
+  }
+
+  val modifiers = Field::class.java.getDeclaredField("accessFlags")
+  modifiers.isAccessible = true
+  modifiers.setInt(field, field.modifiers and Modifier.FINAL.inv())
 }

--- a/packages/expo/CHANGELOG.md
+++ b/packages/expo/CHANGELOG.md
@@ -8,6 +8,8 @@
 
 ### 🐛 Bug fixes
 
+- [Android] Fixed crash from accessing the `accessFlags` attribute on Android API 22. ([#25248](https://github.com/expo/expo/pull/25248) by [@kudo](https://github.com/kudo))
+
 ### 💡 Others
 
 ## 49.0.16 — 2023-10-20

--- a/packages/expo/android/src/main/java/expo/modules/ReactActivityDelegateWrapper.kt
+++ b/packages/expo/android/src/main/java/expo/modules/ReactActivityDelegateWrapper.kt
@@ -3,6 +3,7 @@ package expo.modules
 import android.app.Activity
 import android.content.Context
 import android.content.Intent
+import android.os.Build
 import android.os.Bundle
 import android.view.KeyEvent
 import android.view.ViewGroup
@@ -123,9 +124,7 @@ class ReactActivityDelegateWrapper(
     if (newDelegate != null && newDelegate != this) {
       val mDelegateField = ReactActivity::class.java.getDeclaredField("mDelegate")
       mDelegateField.isAccessible = true
-      val modifiers = Field::class.java.getDeclaredField("accessFlags")
-      modifiers.isAccessible = true
-      modifiers.setInt(mDelegateField, mDelegateField.modifiers and Modifier.FINAL.inv())
+      removeFinalModifier(mDelegateField)
       mDelegateField.set(activity, newDelegate)
       delegate = newDelegate
 
@@ -298,6 +297,25 @@ class ReactActivityDelegateWrapper(
       methodMap[name] = method
     }
     return method!!.invoke(delegate, *args) as T
+  }
+
+  /**
+   * Remove the Java `final` modifier from the field
+   */
+  @Suppress("DiscouragedPrivateApi")
+  private fun removeFinalModifier(field: Field) {
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
+      val artField = Field::class.java.getDeclaredField("artField")
+      artField.isAccessible = true
+      val modifiers = Class.forName("java.lang.reflect.ArtField").getDeclaredField("accessFlags")
+      modifiers.isAccessible = true
+      modifiers.setInt(artField.get(field), field.modifiers and Modifier.FINAL.inv())
+      return
+    }
+
+    val modifiers = Field::class.java.getDeclaredField("accessFlags")
+    modifiers.isAccessible = true
+    modifiers.setInt(field, field.modifiers and Modifier.FINAL.inv())
   }
 
   //endregion

